lib/iris/experimental/ugrid/mesh.py

```diff
@@ -1962,6 +1962,292 @@ def topology_dimension(self):
         return self._metadata_manager.topology_dimension
 
 
+class MeshIndexSet(Mesh):
+    def __init__(self, mesh, location, indices):
+        self.super_mesh = mesh
+        self.location = location
+        self.indices = indices
+
+        self._metadata_manager = metadata_manager_factory(MeshMetadata)
+
+        # topology_dimension is read-only, so assign directly to the metadata manager
+        self._metadata_manager.topology_dimension = mesh.topology_dimension
+
+        self.node_dimension = mesh.node_dimension
+        self.edge_dimension = mesh.edge_dimension
+        self.face_dimension = mesh.face_dimension
+
+        # assign the metadata to the metadata manager
+        self.standard_name = mesh.standard_name
+        self.long_name = mesh.long_name
+        self.var_name = mesh.var_name
+        self.units = mesh.units
+        self.attributes = mesh.attributes
+
+        self._coord_manager = _MeshIndexCoordinateManager(mesh, location, indices)
+        self._connectivity_manager = _MeshIndexConnectivityManager(
+            mesh, location, indices
+        )
+
+    def __eq__(self, other):
+        # TBD: this is a minimalist implementation and requires to be revisited
+        return id(self) == id(other)
+
+    def __ne__(self, other):
+        # TBD: this is a minimalist implementation and requires to be revisited
+        return id(self) != id(other)
+
+
+class _MeshIndexManager:
```

Comment:
I'm working on an idea. Mimicking a whole Connectivity or Coord manager leads to a lot of 'coupling' - we are vulnerable to changes in the signature of the original managers. This is particularly relevant given this:

https://github.com/SciTools/iris/blob/8a15e4a839cca463ef8da25ce3b431689e3b024b/lib/iris/mesh/components.py#L595-L605

### Instead

I think we can create a `dict` subclass to replace the `_members` attribute that belongs to the standard manager classes. Something like the below.

The indexing logic you've already written should stay, including the two subclasses specialised for Connectivities and Coords. But rather than mirroring the various properties such as `node_x` and `edge_node_connectivity`, we only need to specialise `__getitem__()`, `items()`, and any other such routine to appropriately index the requested Connectivities/Coords before they are returned.

Finally: `MeshIndexSet` should inherit the standard creation of a Connectivity and Coord manager, then after creation should replace their `_members` attribute with the new dictionary subclass.

This should make it much easier to adapt to future API changes without too many changes to `MeshIndexSet`.

```python
def _readonly(self, *args, **kwargs):
    message = (
        "Modification of MeshIndexSet is forbidden - this is only "
        f"a view onto an original Mesh: id={id(self.mesh)}."
    )
    raise RuntimeError(message)

class MembersDict(dict):
    __setitem__ = _readonly
    __delitem__ = _readonly
    pop = _readonly
    popitem = _readonly
    clear = _readonly
    update = _readonly
    setdefault = _readonly

    mesh: Mesh
    location: str
    indices: np.typing.ArrayLike

    def __init__(self, seq=None, **kwargs):
        for kwarg in ["mesh", "location", "indices"]:
            setattr(self, kwarg, kwargs.pop(kwarg))
        super().__init__(seq, **kwargs)
```

Reply:
A bit different to what I suggested, but a1d38c4a2adc4336c29d24666f5ad663d6c1279d is what I ended up with.